roles/openshift_hosted/tasks/router/router.yml

```diff
@@ -11,69 +11,34 @@
-    - "{{ (openshift_hosted_router_certificate | default({'keyfile':none})).keyfile }}"
-    - "{{ (openshift_hosted_router_certificate | default({'cafile':none})).cafile }}"
-
-  - name: Persist certificate contents
-    openshift_facts:
-      role: hosted
-      openshift_env:
-        openshift_hosted_router_certificate_contents: "{% for certificate in openshift_router_certificate_output.results -%}{{ certificate.content | b64decode }}{% endfor -%}"
-
-  - name: Create PEM certificate
-    copy:
-      content: "{{ openshift.hosted.router.certificate.contents }}"
-      dest: "{{ openshift_master_config_dir }}/openshift-router.pem"
-      mode: 0600
-
-  when: openshift_hosted_router_certificate is defined
+- name: Get the certificate contents for router
+  copy:
+    backup: True
+    dest: "/etc/origin/master/{{ item | basename }}"
+    src: "{{ item }}"
+  with_items: "{{ openshift_hosted_routers | oo_collect(attribute='certificates') |
+                  oo_select_keys_from_list(['keyfile', 'certfile', 'cafile']) }}"
 
 - name: Create OpenShift router
   oc_adm_router:
-    name: "{{ openshift.hosted.router.name | default('router') }}"
-    replicas: "{{ replicas }}"
-    namespace: "{{ openshift.hosted.router.namespace | default('default') }}"
+    name: "{{ item.name }}"
+    replicas: "{{ item.replicas }}"
+    namespace: "{{ item.namespace | default('default') }}"
     # This option is not yet implemented
     # force_subdomain: "{{ openshift.hosted.router.force_subdomain | default(none) }}"
-    service_account: router
-    selector: "{{ openshift.hosted.router.selector | default(none) }}"
-    images: "{{ openshift.hosted.router.registryurl | default(none) }}"
-    default_cert: "{{ openshift_hosted_router_certificate is defined | default(false) | ternary(openshift_master_config_dir + '/openshift-router.pem', omit) }}"
-    # These edits are being specified only to prevent 'changed' on rerun
-    edits:
-    - key: spec.strategy.rollingParams.intervalSeconds
-      value: 1
-      action: put
-    - key: spec.strategy.rollingParams.updatePeriodSeconds
-      value: 1
-      action: put
-    - key: spec.strategy.activeDeadlineSeconds
-      value: 21600
-      action: put
+    service_account: "{{ item.serviceaccount | default('router') }}"
+    selector: "{{ item.selector | default(none) }}"
+    images: "{{ item.images | default(omit) }}"
+    cert_file: "{{ ('/etc/origin/master/' ~ (item.certificates.certfile | basename)) if 'certfile' in item.certificates else omit }}"
+    key_file: "{{ ('/etc/origin/master/' ~ (item.certificates.keyfile | basename)) if 'keyfile' in item.certificates else omit }}"
+    cacert_file: "{{ ('/etc/origin/master/' ~ (item.certificates.cafile | basename)) if 'cafile' in item.certificates else omit }}"
+    edits: "{{ openshift_hosted_router_edits | union(item.edits)  }}"
+    ports: "{{ item.ports }}"
+    stats_port: "{{ item.stats_port }}"
```

Comment:
@kwoodson @abutcher port conflicts may not be an issue for many users, since I think most cases of sharding (outside of Ops) has been for using different routers for different security zones. In that case, the routers are hosted on different sets of nodes, so the port conflicts are not an issue.

We could document that the defaults are the same and need to be overridden explicitly if the sharded routers will be co-located on the same sets of nodes.